Add S3 sink documentation (#4340)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/sinks/s3.md

---
layout: default
title: s3
parent: Sinks
grand_parent: Pipelines
nav_order: 55
---

# s3

The `s3` sink sends records to an Amazon S3 bucket using the S3 client.

## Usage

The following example creates a pipeline configured with an s3 sink. It contains additional options to customize the event and size thresholds for which the pipeline sends record events, and sets the codec type `ndjson`.

```
pipeline:
  ...
  sink:
    - s3:
        aws:
          region: us-east-1
          sts_role_arn: arn:aws:iam::123456789012:role/Data-Prepper
          sts_header_overrides:
        max_retries: 5
        bucket:
          name: bucket_name
          object_key:
            path_prefix: my-elb/%{yyyy}/%{MM}/%{dd}/
        threshold:
          event_count: 2000
          maximum_size: 50mb
          event_collect_timeout: 15s
        codec:
          ndjson:
        buffer_type: in_memory
```

> 

Reviewer: We are missing a top-level configuration section. We should add that, and it should indicate what options are available. I have a few other comments about other items that should get moved up there.

See the S3 source for the current pattern to use.

> ## AWS configuration options

Author: ```suggestion
## Configuration 
```

> Use the following options when customizing the AWS section of the `s3` sink.

Author: ```suggestion
Use the following options when customizing the section of the `s3` sink.
```

Option | Required | Type | Description
:--- | :--- | :--- | :---

> `bucket` | Yes | String | The object from which to retrieve and store data. The `name` must match the name of your object store.

Reviewer 2: ```suggestion
`bucket` | Yes | String | The object from which the data is retrieved and then stored. The `name` must match the name of your object store.
```

> `bucket` | Yes | String | The object from which to retrieve and store data. The `name` must match the name of your object store.

Reviewer: This is in the top-level configuration. That is, it is not under `aws:`.

`region` | No | String | The AWS region to use when connecting to S3. Defaults to the [standard SDK behavior to determine the region](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html).

> `sts_role_arn` | No | String | The [AWS Security Token Service](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) (AWS STS) role that the S3 sink assumes when sending request to S3. Default to the [standard SDK behavior for credentials](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html). 

Reviewer 2: ```suggestion
`sts_role_arn` | No | String | The [AWS Security Token Service](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) (AWS STS) role that the S3 sink assumes when sending a request to S3. Defaults to the [standard SDK behavior for credentials](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html). 
```

`sts_external_id` | No | String | The external ID to attach to AssumeRole requests from AWS STS.

> `max_retries` | No | Integer | The maximum number of times a single request should retry when ingesting data to S3. Defaults to `5`.

Reviewer: This is in the top-level configuration. That is, it is not under `aws:`.

> `object_key` | No | Sets the `path_prefix` and the `file_pattern` of the object store. Defaults to the S3 object `events-%{yyyy-MM-dd'T'hh-mm-ss}` found inside the root directory of the bucket.

Reviewer: This is in the top-level configuration. That is, it is not under `aws:`.

## Threshold configuration options

Use the following options to set ingestion thresholds for the `s3` sink.

Option | Required | Type | Description
:--- | :--- | :--- | :---
`event_count` | Yes | Integer | The maximum number of events the S3 bucket can ingest. 

> `maximum_size` | No | String | The maximum count or size of bytes that the S3 bucket can ingest. Defaults to `50mb`.

Reviewer: This value is required.

> `maximum_size` | No | String | The maximum count or size of bytes that the S3 bucket can ingest. Defaults to `50mb`.

Author: ```suggestion
`maximum_size` | Yes | String | The maximum count or size of bytes that the S3 bucket can ingest. Defaults to `50mb`.
```

`event_collect_timeout` | Yes | String | Sets how long events are collected before ingestion. All values are strings the represent duration, either an ISO_8601 notation string such as `PT20.345S`, or a simple notation such `60s` or `1500ms`.

> ## buffer_type

Reviewer: Our general pattern in the documentation is to only create a sub-section in the documentation when there is a sub-section in the YAML.

This is going against that pattern. So this makes it seem like it follows underneath.

e.g.

```
buffer_type:
  local_file: ???
```

I recommend that you move this into a top-level configuration section.
